pageserver/src/storage_sync/index.rs

```diff
@@ -222,62 +222,142 @@ pub struct RemoteTimeline {
+                            .layer_metadata
+                            .get(path)
+                            .unwrap_or(default_metadata)
+                            .into(),
+                    )
+                })
+                .collect(),
+            missing_layers: index_part
+                .missing_layers
+                .iter()
+                .map(|path| {
+                    (
+                        path.as_path(timeline_path),
+                        index_part
+                            .layer_metadata
+                            .get(path)
+                            .unwrap_or(default_metadata)
+                            .into(),
+                    )
+                })
+                .collect(),
             metadata,
             awaits_download: false,
         })
     }
 }
 
+/// Metadata gathered for each of the layer files.
+///
+/// Fields have to be `Option`s because remote [`IndexPart`]'s can be from different version, which
+/// might have less or more metadata depending if upgrading or rolling back an upgrade.
+#[derive(Debug, Clone, Default, PartialEq, Eq, PartialOrd, Ord)]
+pub struct LayerFileMetadata {
+    file_size: Option<u64>,
+}
+
+impl From<&'_ IndexLayerMetadata> for LayerFileMetadata {
+    fn from(other: &IndexLayerMetadata) -> Self {
+        LayerFileMetadata {
+            file_size: other.file_size,
+        }
+    }
+}
+
+impl LayerFileMetadata {
+    // TODO: probably best to go ahead with builder pattern?
+    pub fn new(file_size: u64) -> Self {
+        LayerFileMetadata {
+            file_size: Some(file_size),
+        }
+    }
+
+    pub fn for_collected_file(p: &std::path::Path) -> anyhow::Result<Self> {
+        let m = std::fs::metadata(p)
+            .with_context(|| format!("Failed to read layer file metadata: {p:?}"))?;
+        let file_size = Some(m.len());
+        Ok(LayerFileMetadata { file_size })
+    }
+}
+
 /// Part of the remote index, corresponding to a certain timeline.
 /// Contains the data about all files in the timeline, present remotely and its metadata.
 #[serde_as]
 #[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
 pub struct IndexPart {
+    #[serde(default)]
+    version: usize,
+
     timeline_layers: HashSet<RelativePath>,
     /// Currently is not really used in pageserver,
     /// present to manually keep track of the layer files that pageserver might never retrieve.
     ///
     /// Such "holes" might appear if any upload task was evicted on an error threshold:
     /// the this layer will only be rescheduled for upload on pageserver restart.
     missing_layers: HashSet<RelativePath>,
+
+    /// Per layer file metadata, which can be present for a present or missing layer file.
+    ///
+    /// Older versions of `IndexPart` will not have this value or have only a part of metadata that
+    /// latest version stores.
+    #[serde(default)]
+    layer_metadata: HashMap<RelativePath, IndexLayerMetadata>,
```

Comment:
Curious, if we could somehow unite that with `timeline_layers: HashSet<RelativePath>`? 

From what I understand now, it's either an old `index_part.json`, where we have no layer metadata at all, or a new json file, where we have all layers with metadata.

So, we could add some enum here, similar to what you've had with `V1` in the previous, closed PR on the topic.
```rust
pub struct IndexPartJson {
    Legacy {
        //... other fields
        // Serde will deserialize old format here automagically
        timeline_layers: HashSet<RelativePath>,
    },
    V1 {
        //... other fields, we can share them in some common struct, if needed, and `flatten` it
        // Serde will deserialize new format here automagically
        timeline_layers: HashMap<RelativePath, IndexLayerMetadata>
    },
}
```

Presumably, we're able to structure the code so that only S3 downloads might have such `Legacy` entries, we can download such entries right between the timeline download and timeline attach (load into pageserver's memory).
This way, we'll keep `IndexPartJson` on the storage sync module border and provide the rest of the code with the migrated `IndexPart` version that contains proper values.

Reply:
Similar concerns are about `LayerFileMetadata` and `IndexLayerMetadata`: if we keep to the "all or nothing" layer metadata logic, `LayerFileMetadata` has to have a non-optional field in it.

Reply:
I am very much open to going back to versioned enum, be it a document version or version from the metadata header. Starting this journey with one was at least quite difficult with the versioned one, I think it will be easier to transition back once the other parts are done. I am unsure how could we be able to roll back a release with this approach though.

Agreed with the lack of optional fields with this approach.

I don't think one can put this as enum with two struct variants. Supporting this kind of structure will AFAIK require some mix of V1 and V2 types -- more if having the common flattened one, but you probably meant the code example just as an illustration.

Reply:
Resolving this thread for: https://github.com/neondatabase/neon/pull/2582#issuecomment-1277472252